cellar_rebalancer/src/allocation.rs

```
@@ -0,0 +1,157 @@
+
+        let fee = Coin {
+            denom: cosmos_gas_price.1,
+            amount: 0u32.into(),
+        };
+
+        match timeout(Duration::from_secs(100), async {
+            loop {
+                // Waiting for new vote period
+                let res = somm_send::query_commit_period(grpc_client).await;
+                if let Ok(val) = res {
+                    if val.vote_period_start == val.current_height {
+                        break val;
+                    }
+                }
+                sleep(Duration::from_secs(1)).await;
+            }
+        })
+        .await
+        {
+            Ok(val) => {
+                // Sending Pre-commits
+                somm_send::send_precommit(
+                    &contact,
+                    delegate_cosmos_address,
+                    cosmos_key,
+                    fee.clone(),
+                    vec![allocation_precommit(&cosmos_key, &allocation, pair_id.to_string()).await],
+                )
+                .await
+                .unwrap();
+            }
+            Err(e) => {
+                println!("Couldn't Send Precommit {:?}", e);
+            }
+        }
+
+        match timeout(Duration::from_secs(100), async {
+            loop {
+                // Checking Pre-commits for validators
+                let res = somm_send::query_allocation_precommits(grpc_client).await;
+                if let Ok(val) = res {
+                    break val;
+                }
+                sleep(Duration::from_secs(1)).await;
+            }
+        })
+        .await
+        {
+            Ok(val) => {
+                // Sending Commits
+                somm_send::send_allocation(
+                    &contact,
+                    delegate_cosmos_address,
+                    cosmos_key,
+                    fee,
+                    vec![allocation],
+                )
+                .await
+                .unwrap();
```

Same deal here, let's propagate an error if we hit one.